internal: add EventTime to timestamp filtering

dgrisonnet opened the pull request:
In the case of a singleton Event generated by `k8s.io/client-go/tools/events` EventRecorder, the Event creation Timestamp is only set inside the EventTime field. Leaving both the Series and the LastTimestamp nil. Thus, we also need to check the EventTime before updating kube_events_total counter.

lilic commented on internal/collector/event.go:
```diff
@@ -100,7 +100,15 @@ func beforeLastEvent(t time.Time, ev *v1.Event) bool {
 	}
 
-	return t.Before(ev.LastTimestamp.Time)
+	if !ev.LastTimestamp.IsZero() {
```
I would like you to add a unit test for this function `beforeLastEvent`, you could easily unit test it with the various edge cases it also might help us find a few other edge cases.

dgrisonnet replied:
@lilic I have added some test scenarios in my last commit, would you mind having another look?

lilic commented on internal/collector/event.go:
```diff
@@ -68,7 +68,7 @@ func (collector *EventCollector) Run(stopCh <-chan struct{}) {
 			// Count only Events created after the exporter starts running.
-			if beforeLastEvent(startRunning, ev) {
+			if beforeLatestEvent(startRunning, ev) {
 				// FIXME: take into account the event count.
```
Is this still a valid TODO?

dgrisonnet replied:
No, I removed it. Thanks for noticing it :)

lilic commented on internal/collector/event_test.go:
```diff
@@ -0,0 +1,123 @@
+)
+
+func TestBeforeLatestEvent(t *testing.T) {
+	now := time.Now()
```
This should not be time.Now() but a fixed time in the past. e.g. Using `time.Date()` https://golang.org/pkg/time/#Date this makes sure there are no flakes and time is fixed.

dgrisonnet commented:
@lilic thank you for the review! :slightly_smiling_face:  I addressed your comments, please have another look.

lilic commented on internal/collector/event_test.go:
```diff
@@ -0,0 +1,123 @@
+		{
+			Desc: "only older time instants",
+			Event: &v1.Event{
+				FirstTimestamp: metav1.NewTime(older),
```
Could we have a case for:
- newer, older, now
- newer, older, older
- etc?

We just want to cover all cases, unless you think these are the only ones that we need to cover?

dgrisonnet replied:
I don't think it is meaningful to cover all cases tbh. There are a lot of scenarios that we could provide to this function but in practice, there will only be 4-5 scenarios.
For this test, I did not provide real Event scenarios as those will most likely be tested in e2e tests. However, the scenario I came with were: "one newer timestamp", "only one newer timestamp field provided", "no newer timestamp", "only one older timestamp field provided". What do you think?

lilic replied:
> function but in practice, there will only be 4-5 scenarios.

I don't see 4-5 scenarios here. It's not urgent to add them, but nice to add have more of them. It's not blocking this PR.

lilic commented on internal/collector/event.go:
```diff
@@ -95,12 +94,25 @@ func (collector *EventCollector) increaseEventsTotal(event *v1.Event, nbNew floa
-		return t.Before(ev.Series.LastObservedTime.Time)
+func beforeLatestEvent(t time.Time, ev *v1.Event) bool {
+	eventTimes := []time.Time{
+		ev.FirstTimestamp.Time,
```
what if event is nil?

dgrisonnet replied:
It should never be nil as the Events are provided by the informer event handler.

dgrisonnet commented:
@lilic I changed the test cases to have real case scenarios and split the different scenarios into sub-tests that are run in parallel. Please have another look :slightly_smiling_face:
